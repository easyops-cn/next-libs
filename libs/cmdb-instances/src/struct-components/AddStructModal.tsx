import React, { ChangeEvent } from "react";
import { DatePicker, Input, InputNumber, Modal, Radio, Select } from "antd";
import { Attribute, Structkey } from "./interfaces";
import { boolOptions } from "../model-attribute-form-control/ModelAttributeFormControl";
import styles from "./index.module.css";
import i18n from "i18next";
import { K, NS_LIBS_CMDB_INSTANCES } from "../i18n/constants";
import moment from "moment";
import { RadioChangeEvent } from "antd/lib/radio";
import { SelectValue } from "antd/lib/select";
import { computeDateFormat } from "../processors";
import _ from "lodash";
import { CodeEditor } from "@next-libs/code-editor-components";

export interface AddStructModalProps {
  structData?: any;
  attribute: Attribute;
  handleStoreFunction: Function; // eslint-disable-line
  handleCancelFunction: Function; // eslint-disable-line
  visible: boolean;
}
export interface AddStructModalState {
  structData?: any;
  // 格式是否匹配数组
  showError?: boolean[];
}

export class AddStructModal extends React.Component<
  AddStructModalProps,
  AddStructModalState
> {
  constructor(props: AddStructModalProps) {
    super(props);
    const { attribute } = props;
    this.state = {
      showError: new Array(attribute.value.struct_define.length).fill(false),
    };
  }
  componentDidUpdate(prevProps: AddStructModalProps) {
    if (prevProps.visible !== this.props.visible) {
      if (this.props.visible) {
        this.setState({ structData: this.props.structData || {} });
      }
    }
  }
  handleValueChange = (e: SelectValue, define: Structkey) => {
    const structData = this.state.structData;
    structData[define.id] = e;
    this.setState({ structData });
  };
  handleInputValueChange = (
    e:
      | RadioChangeEvent
      | ChangeEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLTextAreaElement>,
    define: any
  ) => {
    this.handleValueChange(e.target.value, define);
  };
  handleIpValueChange = (
    e: ChangeEvent<HTMLInputElement>,
    define: any,
    index: number
  ) => {
    const value = e.target.value;
    const { showError } = this.state;
    const matched =
      value === "" ||
      /^((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)\.){3}(25[0-5]|2[0-4]\d|1\d{2}|[1-9]?\d)(\[^\\],;\s]{1,100}\]|)$/.test(
        value
      );
    showError[index] = !matched;
    this.setState({
      showError,
    });
    this.handleValueChange(value, define);
  };
  getEnumForm = (define: any, value: string) => {
    let enumForm;
    const { regex } = define;
    if (regex.length < 6) {
      const RadioGroup = Radio.Group;

      enumForm = (
        <RadioGroup
          onChange={(e) => this.handleInputValueChange(e, define)}
          defaultValue={value}
        >
          {regex.map((item: string) => (
            <Radio value={item} key={item}>
              {item}
            </Radio>
          ))}
        </RadioGroup>
      );
    } else {
      const Option = Select.Option;
      enumForm = (
        <Select
          {...(value !== null ? { defaultValue: value } : {})}
          style={{ width: "100%" }}
          onChange={(e: SelectValue) => this.handleValueChange(e, define)}
        >
          {regex.map((item: string) => (
            <Option value={item} key={item}>
              {item}
            </Option>
          ))}
        </Select>
      );
    }
    return enumForm;
  };
  validateJson = (err: any, index: number) => {
    const { showError } = this.state;
    const error = _.some(err, ["type", "error"]);
    showError[index] = error;
    this.setState({ showError });
  };
  getFormType = (define: Structkey, value: any, index: number) => {
    let formType;
    let defaultValue = value ? value[define.id] : null;
    //如果是json类型，数据有可能是字符串、数组或对象等，需要对数据处理
    if (define.type === "json") {
      defaultValue =
        _.isString(defaultValue) || !defaultValue
          ? defaultValue || ""
          : JSON.stringify(defaultValue, null, 2);
    }
    switch (define.type) {
      case "int": {
        formType = (
          <InputNumber
            defaultValue={defaultValue}
            style={{ width: "100%" }}
            precision={0}
            onChange={(e) => this.handleValueChange(e, define)}
          />
        );
        break;
      }
      case "enum": {
        formType = this.getEnumForm(define, defaultValue);
        break;
      }
      case "enums": {
        formType = (
          <Select
            mode="multiple"
            style={{ width: "100%" }}
            onChange={(e: SelectValue) => this.handleValueChange(e, define)}
            {...(defaultValue !== null ? { defaultValue: defaultValue } : {})}
          >
            {define.regex.map((item: string) => (
              <Select.Option value={item} key={item}>
                {item}
              </Select.Option>
            ))}
          </Select>
        );
        break;
      }
      case "bool": {
        formType = (
          <Radio.Group
            onChange={(e) => this.handleInputValueChange(e, define)}
            defaultValue={defaultValue}
          >
            {boolOptions.map((item) => (
              <Radio value={item.id} key={item.text}>
                {item.text}
              </Radio>
            ))}
          </Radio.Group>
        );
        break;
      }
      case "arr": {
        formType = (
          <Select
            mode="tags"
            style={{ width: "100%" }}
            {...(defaultValue !== null ? { defaultValue: defaultValue } : {})}
            onChange={(e: SelectValue) => this.handleValueChange(e, define)}
          />
        );
        break;
      }
      case "date":
      case "datetime": {
        const formater = computeDateFormat(define.type, defaultValue);
        formType = (
          <DatePicker
            defaultValue={formater.value}
            showTime={define.type === "datetime"}
            onChange={(e) =>
              this.handleValueChange(
                e.format(
                  define.type === "datetime"
                    ? "YYYY-MM-DD HH:mm:ss"
                    : "YYYY-MM-DD"
                ),
                define
              )
            }
            format={formater.format}
          />
        );
        break;
      }
      case "ip": {
        formType = (
          <div>
            <Input
              defaultValue={defaultValue}
              onChange={(e) => this.handleIpValueChange(e, define, index)}
            />
            <label
              style={{
                display: this.state.showError[index] ? "block" : "none",
                color: "#fc5043",
              }}
            >
              {i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.NOT_MEET_REGEX}`)}
            </label>
          </div>
        );
        break;
      }
      case "float": {
        formType = (
          <InputNumber
            defaultValue={defaultValue}
            style={{ width: "100%" }}
            onChange={(e) => this.handleValueChange(e, define)}
          />
        );
        break;
      }
      case "json": {
        formType = (
          <div>
            <CodeEditor
              value={defaultValue}
              mode={"json"}
              maxLines={"Infinity"}
              highlightActiveLine={true}
              onChange={(e: any) => this.handleValueChange(e, define)}
              minLines={3}
              showLineNumbers={true}
              showPrintMargin={false}
              onValidate={(err: any) => this.validateJson(err, index)}
            />
            <label
              style={{
                display: this.state.showError[index] ? "block" : "none",
                color: "#fc5043",
              }}
            >
              {i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.NOT_MEET_JSON}`)}
            </label>
          </div>
        );
        break;
      }
      default: {
        formType = (
          <Input
            defaultValue={defaultValue}
            onChange={(e) => this.handleInputValueChange(e, define)}
          />
        );
      }
    }
    return formType;
  };

  // 保存修改
  handleStore = () => {
    this.props.handleStoreFunction(this.state.structData);
  };
  handleCloseModal = () => {
    this.props.handleCancelFunction();
  };
  render() {
    const { attribute, visible, structData } = this.props;
    const structDefine = attribute.value.struct_define;
    return (
      <Modal
        visible={visible}
        onOk={this.handleStore}
        onCancel={this.handleCloseModal}
        okButtonProps={{
          disabled: this.state.showError.includes(true),
          id: "okBtn",
        }}
        okText={i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.CONFIRM}`)}
        cancelText={i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.CANCEL}`)}
        cancelButtonProps={{ id: "cancelBtn" }}
        destroyOnClose={true}
      >
        {structDefine.map((define, index) => {
          return (
            <div className={styles.formItem} key={define.id}>
              <label className={styles.formLabel}>{define.name}：</label>
              {this.getFormType(define, structData, index)}
            </div>
          );
        })}
      </Modal>
    );
  }
}
