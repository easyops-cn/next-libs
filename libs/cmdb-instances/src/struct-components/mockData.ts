import { Attribute } from "./interfaces";
export const structData = {
  arr: ["abc", "def"],
  str: "string",
  int: 1000,
  ip: "0.0.0.0",
  enum: "1",
  date: "2019-05-24",
  datetime: "2019-05-24 12:00:00",
  enums: ["1", "3"],
  float: 100.1,
  json: "666",
  json1: { test: "666" },
};
export const structData2 = {
  arr: ["ppp"],
  str: "string2",
  int: 100,
  ip: "10.0.0.0",
  enum: "5",
  date: "2019-05-23",
  datetime: "2019-05-24 23:00:00",
  enums: ["1", "3", "5"],
  float: 100.1,
};
export const structList = [structData, structData2];
export const attribute: Attribute = {
  name: "结构体",
  id: "struct",
  value: {
    type: "structs",
    struct_define: [
      {
        id: "arr",
        name: "数组",
        type: "arr",
      },
      {
        id: "str",
        name: "字符串",
        type: "str",
      },
      {
        id: "int",
        name: "数组",
        type: "int",
      },
      {
        id: "ip",
        name: "IP地址",
        type: "ip",
      },
      {
        id: "enum",
        name: "枚举",
        type: "enum",
        regex: ["1", "3", "5", "7", "9"],
      },
      {
        id: "date",
        name: "日期",
        type: "date",
      },
      {
        id: "datetime",
        name: "时间",
        type: "datetime",
      },
      {
        id: "enums",
        name: "枚举多选",
        type: "enums",
        regex: ["1", "3", "5", "7", "9"],
      },
      {
        id: "float",
        name: "浮点型",
        type: "float",
      },
      {
        id: "json",
        name: "json",
        type: "json",
      },
      {
        id: "json1",
        name: "json1",
        type: "json",
      },
    ],
  },
};
