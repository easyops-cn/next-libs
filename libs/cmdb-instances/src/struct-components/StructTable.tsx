import React from "react";
import { DeleteOutlined, EditOutlined } from "@ant-design/icons";
import { Button, Modal, Table } from "antd";
import { Attribute, Structkey } from "./interfaces";
import { isEmpty, isObject } from "lodash";
import { AddStructModal } from "./AddStructModal";
import { SizeType } from "antd/lib/config-provider/SizeContext";
import styles from "./index.module.css";
import i18n from "i18next";
import { K, NS_LIBS_CMDB_INSTANCES } from "../i18n/constants";
import { Typography } from "antd";
export interface StructTableProps {
  attribute: Attribute;
  structData: any;
  isEditable?: boolean;
  // 是否单结构体
  isLegacy?: boolean;
  size?: SizeType;
  handleStoreFunction?(data: any): void;
}
export interface StructTableState {
  structData: any;
  currentIndex: number;
  showEditModal: boolean;
}
export class StructTable extends React.Component<
  StructTableProps,
  StructTableState
> {
  constructor(props: StructTableProps) {
    super(props);
    this.state = {
      structData: props.structData,
      currentIndex: 0,
      showEditModal: false,
    };
  }
  getColumns(defines: Structkey[]) {
    const columns = defines.map((item: Structkey) => ({
      title: item.name,
      className: styles.structTableTd,
      dataIndex: item.id,
      render: (text: any, _record: any, _index: number) => {
        let copyText: string;
        let showText: string;
        switch (item.type) {
          case "bool":
            return String(text);
          case "enums":
          case "arr":
            return text?.join?.("; ");
          case "json":
            copyText = isObject(text) ? JSON.stringify(text, null, 2) : text;
            showText = isObject(text) ? JSON.stringify(text) : text;
            if (copyText) {
              return (
                <Typography.Paragraph
                  copyable={{
                    tooltips: [
                      i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.COPY}`),
                      i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.COPY_SUCCESS}`),
                    ],
                    text: copyText,
                  }}
                >
                  {showText}
                </Typography.Paragraph>
              );
            }
            return showText;
          default:
            return text;
        }
      },
    }));
    if (this.props.isEditable) {
      columns.push({
        title: i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.OPERATION}`),
        className: styles.structTableTd,
        dataIndex: "operation",
        render: (_text: string, record: any, index: number): any => {
          return this.renderOperation(record, index);
        },
      });
    }
    return columns;
  }
  openConfirmModal(index: number) {
    const confirm = Modal.confirm;

    confirm({
      title: i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.DELETE_STRUCT_CONFIRM_MSG}`),
      okText: i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.CONFIRM}`),
      okType: "danger",
      cancelText: i18n.t(`${NS_LIBS_CMDB_INSTANCES}:${K.CANCEL}`),
      onOk: () => {
        this.remove(index);
      },
    });
  }
  remove(index: number) {
    let { structData, isLegacy } = this.props;
    if (isLegacy) {
      structData = {};
    } else {
      structData.splice(index, 1);
    }
    this.props.handleStoreFunction(structData);
  }
  handleEditStruct = (formData: any, index: number) => {
    let { structData, isLegacy } = this.props;
    if (isLegacy) {
      structData = [structData];
    }
    structData[index] = formData;
    this.props.handleStoreFunction(structData);
    this.handleCloseModal();
  };
  handleOpenEditModal = (index: number) => {
    this.setState({
      showEditModal: true,
      currentIndex: index,
    });
  };
  handleCloseModal = () => {
    this.setState({ showEditModal: false });
  };
  renderOperation = (record: any, index: number) => {
    const { attribute, structData } = this.props;
    const { currentIndex } = this.state;
    return (
      <div>
        <Button
          type="link"
          icon={<EditOutlined />}
          onClick={() => this.handleOpenEditModal(index)}
        />
        <Button
          type="link"
          icon={<DeleteOutlined style={{ color: "var(--theme-red-color)" }} />}
          onClick={() => {
            this.openConfirmModal(index);
          }}
        />
        <AddStructModal
          attribute={attribute}
          visible={this.state.showEditModal && currentIndex === index}
          structData={record}
          handleStoreFunction={(formData: any) =>
            this.handleEditStruct(formData, index)
          }
          handleCancelFunction={this.handleCloseModal}
        />
      </div>
    );
  };
  render() {
    const { attribute, isLegacy, structData } = this.props;
    const structDefine = attribute.value.struct_define;
    const columns = this.getColumns(structDefine);
    // 单结构体数据是对象，显示时要转换为数组
    const dataSource = isEmpty(structData)
      ? []
      : isLegacy
      ? [structData]
      : structData;
    return (
      <div style={{ overflowX: "hidden" }}>
        <Table
          scroll={{ x: "max-content" }}
          columns={columns}
          dataSource={dataSource}
          pagination={false}
          size={this.props.size}
        />
      </div>
    );
  }
}
