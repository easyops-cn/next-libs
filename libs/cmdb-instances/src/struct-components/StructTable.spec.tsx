import React from "react";
import { shallow } from "enzyme";
import { StructTable } from "./StructTable";
import { AddStructModal } from "./AddStructModal";
import { attribute, structData, structData2 } from "./mockData";
import { Button, Modal } from "antd";

const spyOnModalConfirm = jest.spyOn(Modal, "confirm");

describe("StructTable", () => {
  const handleStoreFunction = jest.fn();
  const structsProps = {
    attribute,
    structData: [structData, structData2],
    isEditable: true,
    isLegacy: false,
    handleStoreFunction,
  };
  const structProps = {
    attribute,
    structData: structData,
    isEditable: true,
    isLegacy: true,
    handleStoreFunction,
  };
  const structsWrapper = shallow(<StructTable {...structsProps} />);
  const structsInstance = structsWrapper.instance() as StructTable;
  const structWrapper = shallow(<StructTable {...structProps} />);
  const structInstance = structWrapper.instance() as StructTable;
  const structsOperationWrapper = shallow(
    structsInstance.renderOperation(structData, 0)
  );
  const structOperationWrapper = shallow(
    structInstance.renderOperation(structData, 0)
  );
  const formData = { str: "newString" };
  it("should render", () => {
    expect(structsWrapper.find("Table").length).toBe(1);
    expect(structsWrapper).toBeTruthy();
    expect(structWrapper).toBeTruthy();
    expect(structsOperationWrapper).toBeTruthy();
    expect(structOperationWrapper).toBeTruthy();
  });
  // 打开结构体编辑弹窗
  it("should call handleOpenEditModal function", () => {
    structsOperationWrapper.find(Button).at(0).prop("onClick")(
      expect.anything()
    );
    expect(structsInstance.state.showEditModal).toEqual(true);
    expect(structsInstance.state.currentIndex).toEqual(0);
  });
  // 打开删除确认弹窗
  it("should call openConfirmModal function", () => {
    structsOperationWrapper.find(Button).at(1).prop("onClick")(
      expect.anything()
    );
    expect(spyOnModalConfirm).toBeCalled();
    expect(spyOnModalConfirm.mock.calls[
      spyOnModalConfirm.mock.calls.length - 1
    ][0].title).toEqual("libs-cmdb-instances:DELETE_STRUCT_CONFIRM_MSG");
  });
  //编辑结构体
  it("should call the edit structs function", () => {
    const spy = jest.spyOn(structsProps, "handleStoreFunction");
    structsOperationWrapper
      .find(AddStructModal)
      .at(0)
      .prop("handleStoreFunction")(formData, 0);
    expect(spy).toBeCalledWith([formData, structData2]);
  });
  // 删除结构体
  it("should call with a struct array", () => {
    const spy = jest.spyOn(structsProps, "handleStoreFunction");
    structsInstance.remove(0);
    expect(spy).toBeCalledWith([structData2]);
  });
  //编辑单结构体
  it("should call the edit struct function", () => {
    const spy = jest.spyOn(structProps, "handleStoreFunction");
    structOperationWrapper
      .find(AddStructModal)
      .at(0)
      .prop("handleStoreFunction")(formData, 0);
    expect(spy).toBeCalledWith([formData]);
  });
  // 删除单结构体
  it("should call with a struct object", () => {
    const spy = jest.spyOn(structsProps, "handleStoreFunction");
    structInstance.remove(0);
    expect(spy).toBeCalledWith({});
  });
  //点击删除确认弹窗的确定按钮
  it("should call the remove function when click Ok button of confirm modal", () => {
    structsOperationWrapper.find(Button).at(1).prop("onClick")(
      expect.anything()
    );
    const spy = jest.spyOn(structsProps, "handleStoreFunction");
    spyOnModalConfirm.mock.calls[
      spyOnModalConfirm.mock.calls.length - 1
    ][0].onOk();
    expect(spy).toBeCalled();
  });
});
