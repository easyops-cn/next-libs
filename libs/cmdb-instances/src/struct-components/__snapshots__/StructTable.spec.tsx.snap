// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StructTable should render 1`] = `
<div
  style={
    Object {
      "overflowX": "hidden",
    }
  }
>
  <Table
    columns={
      Array [
        Object {
          "className": "structTableTd",
          "dataIndex": "arr",
          "render": [Function],
          "title": "数组",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "str",
          "render": [Function],
          "title": "字符串",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "int",
          "render": [Function],
          "title": "数组",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "ip",
          "render": [Function],
          "title": "IP地址",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "enum",
          "render": [Function],
          "title": "枚举",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "date",
          "render": [Function],
          "title": "日期",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "datetime",
          "render": [Function],
          "title": "时间",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "enums",
          "render": [Function],
          "title": "枚举多选",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "operation",
          "render": [Function],
          "title": "操作",
        },
      ]
    }
    dataSource={
      Array [
        Object {
          "arr": Array [
            "abc",
            "def",
          ],
          "date": "2019-05-24",
          "datetime": "2019-05-24 12:00:00",
          "enum": "1",
          "enums": Array [
            "1",
            "3",
          ],
          "int": 1000,
          "ip": "0.0.0.0",
          "str": "string",
        },
        Object {
          "arr": Array [
            "ppp",
          ],
          "date": "2019-05-23",
          "datetime": "2019-05-24 23:00:00",
          "enum": "5",
          "enums": Array [
            "1",
            "3",
            "5",
          ],
          "int": 100,
          "ip": "10.0.0.0",
          "str": "string2",
        },
      ]
    }
    pagination={false}
    rowKey="key"
    scroll={
      Object {
        "x": "max-content",
      }
    }
  />
</div>
`;

exports[`StructTable should render 2`] = `
<div
  style={
    Object {
      "overflowX": "hidden",
    }
  }
>
  <Table
    columns={
      Array [
        Object {
          "className": "structTableTd",
          "dataIndex": "arr",
          "render": [Function],
          "title": "数组",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "str",
          "render": [Function],
          "title": "字符串",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "int",
          "render": [Function],
          "title": "数组",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "ip",
          "render": [Function],
          "title": "IP地址",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "enum",
          "render": [Function],
          "title": "枚举",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "date",
          "render": [Function],
          "title": "日期",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "datetime",
          "render": [Function],
          "title": "时间",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "enums",
          "render": [Function],
          "title": "枚举多选",
        },
        Object {
          "className": "structTableTd",
          "dataIndex": "operation",
          "render": [Function],
          "title": "操作",
        },
      ]
    }
    dataSource={
      Array [
        Object {
          "arr": Array [
            "abc",
            "def",
          ],
          "date": "2019-05-24",
          "datetime": "2019-05-24 12:00:00",
          "enum": "1",
          "enums": Array [
            "1",
            "3",
          ],
          "int": 1000,
          "ip": "0.0.0.0",
          "str": "string",
        },
      ]
    }
    pagination={false}
    rowKey="key"
    scroll={
      Object {
        "x": "max-content",
      }
    }
  />
</div>
`;

exports[`StructTable should render 3`] = `
<div>
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    icon={<ForwardRef(EditOutlined) />}
    loading={false}
    onClick={[Function]}
    type="link"
  />
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    icon={<ForwardRef(DeleteOutlined) />}
    loading={false}
    onClick={[Function]}
    type="link"
  />
  <AddStructModal
    attribute={
      Object {
        "id": "struct",
        "name": "结构体",
        "value": Object {
          "struct_define": Array [
            Object {
              "id": "arr",
              "name": "数组",
              "type": "arr",
            },
            Object {
              "id": "str",
              "name": "字符串",
              "type": "str",
            },
            Object {
              "id": "int",
              "name": "数组",
              "type": "int",
            },
            Object {
              "id": "ip",
              "name": "IP地址",
              "type": "ip",
            },
            Object {
              "id": "enum",
              "name": "枚举",
              "regex": Array [
                "1",
                "3",
                "5",
                "7",
                "9",
              ],
              "type": "enum",
            },
            Object {
              "id": "date",
              "name": "日期",
              "type": "date",
            },
            Object {
              "id": "datetime",
              "name": "时间",
              "type": "datetime",
            },
            Object {
              "id": "enums",
              "name": "枚举多选",
              "regex": Array [
                "1",
                "3",
                "5",
                "7",
                "9",
              ],
              "type": "enums",
            },
          ],
          "type": "structs",
        },
      }
    }
    handleCancelFunction={[Function]}
    handleStoreFunction={[Function]}
    structData={
      Object {
        "arr": Array [
          "abc",
          "def",
        ],
        "date": "2019-05-24",
        "datetime": "2019-05-24 12:00:00",
        "enum": "1",
        "enums": Array [
          "1",
          "3",
        ],
        "int": 1000,
        "ip": "0.0.0.0",
        "str": "string",
      }
    }
    visible={false}
  />
</div>
`;

exports[`StructTable should render 4`] = `
<div>
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    icon={<ForwardRef(EditOutlined) />}
    loading={false}
    onClick={[Function]}
    type="link"
  />
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    icon={<ForwardRef(DeleteOutlined) />}
    loading={false}
    onClick={[Function]}
    type="link"
  />
  <AddStructModal
    attribute={
      Object {
        "id": "struct",
        "name": "结构体",
        "value": Object {
          "struct_define": Array [
            Object {
              "id": "arr",
              "name": "数组",
              "type": "arr",
            },
            Object {
              "id": "str",
              "name": "字符串",
              "type": "str",
            },
            Object {
              "id": "int",
              "name": "数组",
              "type": "int",
            },
            Object {
              "id": "ip",
              "name": "IP地址",
              "type": "ip",
            },
            Object {
              "id": "enum",
              "name": "枚举",
              "regex": Array [
                "1",
                "3",
                "5",
                "7",
                "9",
              ],
              "type": "enum",
            },
            Object {
              "id": "date",
              "name": "日期",
              "type": "date",
            },
            Object {
              "id": "datetime",
              "name": "时间",
              "type": "datetime",
            },
            Object {
              "id": "enums",
              "name": "枚举多选",
              "regex": Array [
                "1",
                "3",
                "5",
                "7",
                "9",
              ],
              "type": "enums",
            },
          ],
          "type": "structs",
        },
      }
    }
    handleCancelFunction={[Function]}
    handleStoreFunction={[Function]}
    structData={
      Object {
        "arr": Array [
          "abc",
          "def",
        ],
        "date": "2019-05-24",
        "datetime": "2019-05-24 12:00:00",
        "enum": "1",
        "enums": Array [
          "1",
          "3",
        ],
        "int": 1000,
        "ip": "0.0.0.0",
        "str": "string",
      }
    }
    visible={false}
  />
</div>
`;
